### PR TITLE
Introduce block_pointer_cache invalidation counter

### LINKED ISSUE #79: Cached pointers in pmemfile struct's can be invalidated by other file operations.
Certain file operations can remove blocks from a file on media, which can result in some pmemfile structs referring to blocks to be invalidated. Currently there is no trivial way to fix it.

### PR DESCRIPTION
**Pull request by GBuella**
Fixes #79

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/301)

### CODE REVIEW

File: tests/posix/pointer_caching/pointer_caching.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		ASSERT_EQ(r, (decltype(r))buffer_r.size()) << COND_ERROR(r);
+		ASSERT_TRUE(memcmp(buffer_r.data(), buffer_w.data(),
+				   buffer_r.size()) == 0);
+	}
+
+	/*
+	 * At this point, the offset associated with f2 points to somewhere
+	 * close to the end of the file, and should read some more of the data
+	 * written via f1, if used.
+	 * Making this other modification via f1 should shange that -- following
```
Comment (GBuella) [inline tests/posix/pointer_caching/pointer_caching.cpp]: Done.

File: src/libpmemfile-posix/write.c [hunk 2]
```diff
@@ -254,28 +318,86 @@ pmemfile_writev(PMEMfilepool *pfp, PMEMfile *file, const pmemfile_iovec_t *iov,
...
+ * | Erroneous scenario:                                                     |
+ * | checking for modification while the vinode is not locked                |
+ * +-------------------------------------------------------------------------+
+ * | lock(file);                                                             |
+ * |                                                                         |
+ * |  if (is_data_modification_indicated(file)) {  ---+                      |
+ * |     block_pointer_cache = NULL;                  |                      |
+ * |  }                                               |                      |
+ * |  Make a local copy of block_pointer_cache.       | The underlying file  |
+ * |                                                  | can modified here,   |
```
Comment (GBuella) [inline src/libpmemfile-posix/write.c]: Done.

File: src/libpmemfile-posix/write.c [hunk 3]
```diff
@@ -254,28 +318,86 @@ pmemfile_writev(PMEMfilepool *pfp, PMEMfile *file, const pmemfile_iovec_t *iov,
...
 	pmemfile_iovec_t element = {.iov_base = (void *)buf, .iov_len = count};
 	return pmemfile_pwritev(pfp, file, &element, 1, offset);
 }
-
+/*
+ * pmemfile_pwritev - writes to a file starting at a position supplied as
+ * argument.
+ * Since this does not require making any modification to the PMEMfile instance,
+ * the corresponding lock is held only for reading some fields from it. There is
+ * no point in time where this function helds locks of both the PMEMfile
```
Comment (GBuella) [inline src/libpmemfile-posix/write.c]: Done.

File: src/libpmemfile-posix/truncate.c [hunk 4]
```diff
@@ -119,6 +119,8 @@ _pmemfile_ftruncate(PMEMfilepool *pfp, struct pmemfile_vinode *vinode,
 
 	os_rwlock_wrlock(&vinode->rwlock);
 
+	vinode->data_modification_counter++;
+	vinode->metadata_modification_counter++;
```
Comment (GBuella) [inline src/libpmemfile-posix/truncate.c]: Done.

File: src/libpmemfile-posix/file.c [hunk 5]
```diff
@@ -444,6 +444,8 @@ _pmemfile_openat(PMEMfilepool *pfp, struct pmemfile_vinode *dir,
 		if (flags & PMEMFILE_O_TRUNC) {
 			os_rwlock_wrlock(&vinode->rwlock);
 
+			vinode->data_modification_counter++;
+			vinode->metadata_modification_counter++;
```
Comment (GBuella) [inline src/libpmemfile-posix/file.c]: Done.

File: src/libpmemfile-posix/fallocate.c [hunk 6]
```diff
@@ -223,6 +223,9 @@ pmemfile_fallocate(PMEMfilepool *pfp, PMEMfile *file, int mode,
 
 	os_rwlock_wrlock(&vinode->rwlock);
 
+	vinode->data_modification_counter++;
+	vinode->metadata_modification_counter++;
+
```
Comment (GBuella) [inline src/libpmemfile-posix/fallocate.c]: Done.

File: src/libpmemfile-posix/read.c [hunk 7]
```diff
@@ -128,87 +127,80 @@ pmemfile_preadv_internal(PMEMfilepool *pfp,
...
-		} TX_ONABORT {
-			LOG(LINF, "can not update inode atime");
-		} TX_END
+	os_rwlock_wrlock(&vinode->rwlock);
 
-		os_rwlock_unlock(&vinode->rwlock);
-	}
+	TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
+		vinode->metadata_modification_counter++;
+		memory_barrier();
```
Comment (GBuella) [inline src/libpmemfile-posix/read.c]: Maybe.
Note: I like showing this explicitly in the source here. This is rather meaningful if/when we use the modification_counters to detect changes without locking the vinode:
```
modifying_routine(vinode)
{
    vinode->metadata_modification_counter++;
    memory_barrier();
    ...perform modifications...
}

reading_routine(vinode)
{
    ...perform reads...
    memory_barrier();
    if (is_modification_indicated(vinode)) {
       lock(vinode)
        ...perform reads...
       unlock(vinode)
    }    
}
```

**Comment by GBuella** [comment]
I narrowed the scope of this patch somewhat.
It only fixes #79, and doesn't try to prepare a more general solution of counters.

File: src/libpmemfile-posix/inode.h [hunk 8]
```diff
@@ -48,6 +48,12 @@ struct pmemfile_vinode {
 	/* read-write lock, also protects inode read/writes */
 	os_rwlock_t rwlock;
 
+	/*
+	 * Counter to keep track of modifications that potentially
+	 * invalidate a block_pointer_cache field in pmemfile_file struct.
+	 */
+	uint64_t block_pointer_invalidation_counter;
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/inode.h]: It seems you are modifying this only in fallocate and truncate.

Why not do it in vinode_allocate_interval / vinode_remove_interval and only when block_array layout changes (any *existing* block_desc changes position)?

File: src/libpmemfile-posix/read.c [hunk 9]
```diff
@@ -260,14 +323,38 @@ pmemfile_preadv(PMEMfilepool *pfp, PMEMfile *file, const pmemfile_iovec_t *iov,
 		return -1;
 	}
 
+	pmemfile_ssize_t ret;
+
 	os_mutex_lock(&file->mutex);
 
+	ret = pmemfile_preadv_args_check(file, iov, iovcnt);
+
+	uint64_t data_mod_count =
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/read.c]: This should probably be renamed.

File: src/libpmemfile-posix/inode.h (hunk 8 above)
Comment (GBuella) [inline src/libpmemfile-posix/inode.h]: Right, It is enough to increment it only in `vinode_remove_interval`

File: src/libpmemfile-posix/read.c (hunk 9 above)
Comment (GBuella) [inline src/libpmemfile-posix/read.c]: Done

**Comment by marcinslusarz** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemfile-posix/write.c, line 246 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KrqDk6BuMX-xdQRy0Ro:-KrqDk6BuMX-xdQRy0Rp:btu5vfb) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/src/libpmemfile-posix/write.c#L246)):*
> ```C
>  * This function expects the PMEMfile instance to be locked while being called.
>  * Since the offset field is used to determine where to read from, and is also
>  * updated after a successfull read operation, the PMEMfile instance can not be
> ```

s/successfull/successful/

---

*[src/libpmemfile-posix/write.c, line 280 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-Krq92dYE0pWyqYUjlgn:-Krq92dZd__gGDoIIfP6:b-q4em8p) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/src/libpmemfile-posix/write.c#L280)):*
> ```C
> 
> 	file->last_block_pointer_invalidation_observed =
> 		file->vinode->block_pointer_invalidation_counter;
> ```

pwritev_internal can't invalidate block pointer, so I think it would be cleaner to move this to the block above where you check that counters match. The same comment applies to readv_under_filelock.

---

*[tests/posix/CMakeLists.txt, line 177 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KrqE2-82c5_oNkvIlry:-KrqE2-82c5_oNkvIlrz:b-7kid11) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/tests/posix/CMakeLists.txt#L177)):*
> ```CMake
> 
> # Reproducing pointer caching issues can be rather tricky, one
> # would need to rely on very specific details of pmemfile-posix,
> ```

s/,//

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by GBuella** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemfile-posix/write.c, line 246 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KrqDk6BuMX-xdQRy0Ro:-Ks00DVjyl7unzis9Sx_:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/src/libpmemfile-posix/write.c#L246)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

s/successfull/successful/
</blockquote></details>

Done.

---

*[src/libpmemfile-posix/write.c, line 280 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-Krq92dYE0pWyqYUjlgn:-Ks00UT-CQGyouO3KFiY:b-g14ijv) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/src/libpmemfile-posix/write.c#L280)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

pwritev_internal can't invalidate block pointer, so I think it would be cleaner to move this to the block above where you check that counters match. The same comment applies to readv_under_filelock.
</blockquote></details>

I think that is less clean, in the sense that one needs to know "pwritev_internal can't invalidate block pointer" in order to do it before pwritev_internal.
Having it after pwritev_internal doesn't depend on that particular knowledge.

But whatever.

Done.

---

*[tests/posix/CMakeLists.txt, line 177 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KrqE2-82c5_oNkvIlry:-Ks018HiWI049EuOwboG:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/cbb63e729f1a09f61ae2495cb4564c02111b4d4f/tests/posix/CMakeLists.txt#L177)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

s/,//
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Reviewed 1 of 15 files at r1, 1 of 5 files at r2, 2 of 11 files at r3, 2 of 5 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301#-:-Ks39zHL79jqCsD1tp-T:bnfp4nl)*

**Comment by krzycz** [comment]
Reviewed 5 of 15 files at r1, 2 of 5 files at r2, 2 of 11 files at r3, 2 of 5 files at r4, 2 of 3 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by sarahjelinek** [comment]
Reviewed 1 of 5 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[src/libpmemfile-posix/read.c, line 238 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KselPaR98uwo3ZOEMUn:-KselPaR98uwo3ZOEMUo:b-s2wb91) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L238)):*
> ```C
> 		file->last_block_pointer_invalidation_observed =
> 			file->vinode->block_pointer_invalidation_counter;
> 	}
> ```

Is it worth keeping the block cache rather than keeping one copy of the block list in the PMEMfile instance?  I am only asking because we have a lot of complicated logic to try to keep both lists straight and I am wondering if the performance benefit is worth the potential race conditions or incorrectness.

---

*[src/libpmemfile-posix/read.c, line 247 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KseJdrUDQLkyczpHz8L:-KseJdrUDQLkyczpHz8M:b-dqh17y) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L247)):*
> ```C
> 
> 
> 	os_rwlock_unlock(&file->vinode->rwlock);
> ```

Where was this lock taken? I don't see a corresponding case where the lock was taken and not released.  Add a comment above it if it has been taken in another function . 

---

*[src/libpmemfile-posix/read.c, line 303 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KseiD5kFQ6THHeVlhHi:-KseiD5kFQ6THHeVlhHj:brsial2) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L303)):*
> ```C
>  * argument.
>  * Unlocks the PMEMfile instance before locking the vinode instance - for an
>  * explanation of this, see the comments on pmemfile_pwritev in the write.c
> ```

Just put the same comments here as are on pmemfile_pwritev() function. Don't make consumer have to look at another file. 

---

*[src/libpmemfile-posix/read.c, line 356 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-Ksem-JBgYIzry6PbC2C:-Ksem-JBgYIzry6PbC2D:bb8j3ua) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L356)):*
> ```C
> 			(size_t)offset, iov, iovcnt);
> 
> 	os_rwlock_unlock(&file->vinode->rwlock);
> ```

Same as above. Where is this lock taken?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by GBuella** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmemfile-posix/read.c, line 238 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KselPaR98uwo3ZOEMUn:-Ksf2lN_tcX4S8xqc590:b-iniq19) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L238)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

Is it worth keeping the block cache rather than keeping one copy of the block list in the PMEMfile instance?  I am only asking because we have a lot of complicated logic to try to keep both lists straight and I am wondering if the performance benefit is worth the potential race conditions or incorrectness.
</blockquote></details>

For one thing, keeping a copy of the block list (if you mean all block metadata) does not resolve anything here -- the metadata copy might still point to actual data that does not belong to the file already. We would still need to check if that cache is valid or not.

Also the block list can be rather large. Large files might have thousands, or hundreds of thousands of blocks.

About performance benefits, keeping this cache around makes it possible to look up blocks for conscutive sequential reads in `O(1)`, instead of doing `O(log n)` lookups in the blocktree.
As I recall, these blocktree lookups appear to be a bottleneck during random reads (when SQLite does a lot of lseek+read operations). I guess without this cache, sequential reads would show similar symptoms.

---

*[src/libpmemfile-posix/read.c, line 247 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KseJdrUDQLkyczpHz8L:-Ksf2KOCNtB64LqFV6UQ:btxipk4) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L247)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

Where was this lock taken? I don't see a corresponding case where the lock was taken and not released.  Add a comment above it if it has been taken in another function . 
</blockquote></details>

It is taken above, in `vinode_rdlock_with_block_tree`
I was hoping the name vinode_rdlock should make it clear, that it takes a readlock on the vinode.
Open for suggestion for a better name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by sarahjelinek** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemfile-posix/read.c, line 247 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KseJdrUDQLkyczpHz8L:-Ksf86dOaV4dW4dLYbNB:b6p67ls) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L247)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

It is taken above, in `vinode_rdlock_with_block_tree`
I was hoping the name vinode_rdlock should make it clear, that it takes a readlock on the vinode.
Open for suggestion for a better name.
</blockquote></details>

You are right.  I saw that before in a previous look at this code. The name is fine. I just forgot :-). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by krzycz** [comment]
Reviewed 1 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by GBuella** [comment]
Review status: 7 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/libpmemfile-posix/read.c, line 303 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/301#-KseiD5kFQ6THHeVlhHi:-KshlMe17I7a-CRbd6fh:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/1adca10290a6f51a881acdc52215601f8529e234/src/libpmemfile-posix/read.c#L303)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

Just put the same comments here as are on pmemfile_pwritev() function. Don't make consumer have to look at another file. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301)*

**Comment by sarahjelinek** [comment]
:lgtm:

---

Review status: 7 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/301#-:-Ksj9UhSFm2al_NwjdjY:bnfp4nl)*

**Pull request merged**